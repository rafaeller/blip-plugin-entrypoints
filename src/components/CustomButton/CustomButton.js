import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import styles from './styles.module.scss';

const VARIATIONS = {
    primary:
        'bn bp-bg-primary-main bp-c-neutral-light-snow hover-bp-bg-primary-dark active-bp-bg-primary-night',
    'primary-trackings':
        'bn bg-transparent bp-c-neutral-dark-rooftop hover-bp-bg-disabled-bg active-bp-c-neutral-light-snow active-bp-bg-neutral-medium-elephant',
    secondary:
        'bn bg-transparent bp-c-neutral-dark-rooftop hover-bp-c-primary-main hover-bp-bg-disabled-bg active-bp-c-primary-main active-bp-bg-hover-light',
    'secondary-dark':
        'bn bg-transparent bp-c-neutral-medium-silver hover-bp-c-neutral-light-snow hover-bp-bg-neutral-dark-rooftop active-bp-c-neutral-light-snow active-bp-bg-neutral-medium-elephant',
    'secondary-delete':
        'bn bg-transparent bp-c-neutral-dark-rooftop hover-bp-bg-disabled-bg active-bp-c-neutral-light-snow active-bp-bg-extend-reds-delete',
    tertiary:
        'ba bg-transparent bp-c-primary-main bp-bc-neutral-medium-wave hover-bp-bc-primary-main active-bp-bg-hover-light',
    delete: 'bn bp-bg-extend-reds-delete bp-c-neutral-light-snow hover-bp-bg-extend-reds-lipstick active-bp-bg-extend-reds-dragon',
    ghost: 'ba bg-transparent bp-c-primary-main bp-bc-primary-main hover-bp-bg-disabled-bg active-bp-bg-hover-light'
};

const CustomButton = ({
    dataTestId,
    text,
    children,
    icon,
    active = false,
    variant = 'secondary',
    size = 'standard',
    type = 'button',
    iconOnly = false,
    disabled = false,
    wrap = false,
    options = {
        width: 'auto',
        colorClass: ''
    },
    onClick = () => {}
}) => {
    const buttonClass = classNames(
        'flex items-center f6 b br3 outline-0',
        styles.button,
        styles[size],
        styles[variant],
        options.colorClass,
        {
            'w-100': options?.width === '100%',
            'w-auto': options?.width === 'auto',
            nowrap: !wrap,
            pointer: !disabled,
            'not-allowed': disabled,
            [styles.active]: active,
            pv2: true,
            ph2: iconOnly,
            ph3: !iconOnly,
            'justify-center': iconOnly,
            [VARIATIONS[variant]]: !options?.colorClass
        }
    );

    const iconSize = {
        short: 'x-small',
        standard: 'medium'
    };

    return (
        <button
            type={type}
            className={buttonClass}
            disabled={disabled}
            onClick={onClick}
            data-testid={dataTestId}
        >
            {icon && (
                <div
                    className={classNames('flex items-center', {
                        mr1: !iconOnly
                    })}
                >
                    <bds-icon name={icon} size={iconSize[size]} />
                </div>
            )}

            {!iconOnly && (text || children)}
        </button>
    );
};

CustomButton.propTypes = {
    dataTestId: PropTypes.string,
    text: PropTypes.string,
    children: PropTypes.any,
    icon: PropTypes.string,
    active: PropTypes.bool,
    variant: PropTypes.string,
    size: PropTypes.string,
    type: PropTypes.string,
    iconOnly: PropTypes.bool,
    disabled: PropTypes.bool,
    wrap: PropTypes.bool,
    options: PropTypes.object,
    onClick: PropTypes.func
};

export default CustomButton;
