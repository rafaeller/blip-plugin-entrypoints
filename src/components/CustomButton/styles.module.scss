@import '~blip-toolkit/dist/scss/main';
@import '~blip-ds/dist/collection/styles/_colors.scss';

.button {
    transition: all 0.5s ease 0s;

    &.short {
        height: 2rem;
    }

    &.standard {
        height: 2.5rem;
    }

    &.active {
        &.primary-trackings {
            background: $color-primary-main;
            color: $color-neutral-light-snow;
        }

        &.primary {
            background: $color-primary-night;
        }

        &.secondary {
            background: $color-hover-light;
            color: $color-primary-main;
        }

        &.secondary-dark {
            background: $color-neutral-medium-elephant;
            color: #fff;
        }

        &.secondary-delete {
            background: $color-extend-reds-delete;
            color: #fff;
        }

        &.tertiary {
            background: $color-hover-light;
        }

        &.delete {
            background: $color-extend-reds-dragon;
        }

        &.ghost {
            background: $color-hover-light;
        }
    }

    &:disabled {
        cursor: not-allowed;

        &.primary {
            background: $color-primary-light;
        }

        &.secondary,
        &.primary-trackings,
        &.secondary-delete,
        &.tertiary,
        &.ghost {
            background: transparent;
        }

        &.delete {
            background: $color-disabled-wrong;
        }
    }
}
