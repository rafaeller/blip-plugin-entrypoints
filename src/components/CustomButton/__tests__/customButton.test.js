import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import CustomButton from '../CustomButton';

describe('CustomButton component', () => {
    const mockHandle = jest.fn();

    afterEach(() => jest.clearAllMocks());

    it('should handle click event', () => {
        render(
            <CustomButton
                text="Mocked Button"
                variant="secondary"
                size="standard"
                onClick={mockHandle}
                dataTestId="mocked-button"
            />
        );
        const cardNode = screen.queryByTestId('mocked-button');
        userEvent.click(cardNode);
        expect(mockHandle).toHaveBeenCalledTimes(1);
    });

    it('should render correctly', () => {
        const { container } = render(
            <CustomButton
                text="Mocked Button"
                variant="secondary"
                size="standard"
            />
        );

        expect(container.firstChild).toMatchSnapshot();
    });
});
