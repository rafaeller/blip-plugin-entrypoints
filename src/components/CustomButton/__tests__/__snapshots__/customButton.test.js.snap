// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomButton component should render correctly 1`] = `
<button
  class="flex items-center f6 b br3 outline-0 button standard secondary w-auto nowrap pointer pv2 ph3 bn bg-transparent bp-c-neutral-dark-rooftop hover-bp-c-primary-main hover-bp-bg-disabled-bg active-bp-c-primary-main active-bp-bg-hover-light"
  type="button"
>
  Mocked Button
</button>
`;
